gateway/session_manager.go

```diff
@@ -38,45 +40,59 @@ type SessionLimiter struct {
 	Gw          *Gateway `json:"-"`
 }
 
+func (l *SessionLimiter) Context() context.Context {
+	return l.Gw.ctx
+}
+
 func (l *SessionLimiter) doRollingWindowWrite(key, rateLimiterKey, rateLimiterSentinelKey string,
 	currentSession *user.SessionState,
 	store storage.Handler,
 	globalConf *config.Config,
 	apiLimit *user.APILimit, dryRun bool) bool {
 
-	var per, rate float64
+	ctx := l.Context()
+
+	var per, cost float64
 
 	if apiLimit != nil { // respect limit on API level
 		per = apiLimit.Per
-		rate = apiLimit.Rate
+		cost = apiLimit.Rate
 	} else {
 		per = currentSession.Per
-		rate = currentSession.Rate
+		cost = currentSession.Rate
 	}
 
 	log.Debug("[RATELIMIT] Inbound raw key is: ", key)
 	log.Debug("[RATELIMIT] Rate limiter key is: ", rateLimiterKey)
 	pipeline := globalConf.EnableNonTransactionalRateLimiter
 
-	var ratePerPeriodNow int
+	ratelimit, err := rate.NewSlidingLog(store, pipeline)
```

Comment:
shall we maintain this as a singleton member in `SessionLimiter`

Reply:
@jeffy-mathew actually no, as the SessionLimiter is not request scoped. We can't ensure a redis.UniversalClient will live forever, redis can go offline and this is where we need to handle the error. Depending on the situation that may mean that the gateway chokes on incoming requests as they each wait on the connection timeouts, if a singleton would be in use.

Added a constructor benchmark, it's 24B/op; essentially 4 pointers. I don't think they even escape to heap but did not verify.